cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/MetadataStoreDataset.java

```diff
@@ -276,4 +344,34 @@ public void deleteAll(MDSKey id) {
     }
     return builder;
   }
+
+  protected MDSKey.Builder getProgramKeyBuilder(String recordType, @Nullable ProgramRunId programRunId) {
+    MDSKey.Builder builder = new MDSKey.Builder().add(recordType);
+    if (programRunId != null) {
+      builder.add(programRunId.getNamespace());
+      builder.add(programRunId.getApplication());
+      builder.add(programRunId.getVersion());
+      builder.add(programRunId.getType().name());
+      builder.add(programRunId.getProgram());
+    }
+    return builder;
+  }
+
+  /**
+   * Returns a ProgramId given the MDS key
+   *
+   * @param key the MDS key to be used
+   * @return ProgramId created from the MDS key
+   */
+  protected ProgramRunId getProgramRunID(MDSKey key, RunRecordMeta recordMeta) {
```

Comment:
Since `MetadataStoreDataset` does not know anything about run ids in `MDSKey`, it would be good to move extracting the run id part outside this class